### ISSUE TITLE
Admin monthly service report download fails with Lambda API

### ISSUE THREAD

**Issue by patheard** [cds-snc/notification-planning#653]
# Describe the bug 

Attempting to download the monthly service notification report as a platform admin fails when the Admin service is using the Lambda API to handle its requests.

The report downloads as expected if the k8s API pod is used.

# [Bug Severity](https://docs.google.com/document/d/1tcD32bjJRaREsMgaUEoXAk400y1dKzzaKR0G5Sr1w20/edit)

SEV2

# To Reproduce

1. Login to an Admin environment using the Lambda API (currently Staging).
1. Go to [Monthly notification statuses for live services](https://staging.notification.cdssandbox.xyz/platform-admin/reports/notifications-sent-by-service) and select dates.
1. Select "Download report".
1. A technical error screen is displayed.

# Expected behavior

A `.csv` report is downloaded.

# Impact

## Impact on Notify team: 
Unable to download the service notification reports using Lambda API, need to use kn8 API. We want to be fully on Lambda API

# Tasks
- [x] Timeboxed investigation for 1-2 days 
- [x] Dev documents investigation so it can be picked up by other devs if needed

# Additional context

Since we switched the lambda handler library from [awsgi](https://github.com/slank/awsgi) to [apig_wsgi](https://github.com/adamchainz/apig-wsgi), the received JSON response changed from a `list` datatype to a `dict`. More specifically, the Lambda API and k8s API are returning their JSON response in a different format:

## Mismatched JSON outputs between k8s and lambda API

**Lambda API**

```json
// Array of objects
[
    {
        "count_delivered": 22075,
        "count_permanent_failure": 155,
        "count_sending": 0,
        "count_sent": 0,
        "count_technical_failure": 0,
        "count_temporary_failure": 30,
        "date_created": "Sun, 01 Mar 2020 00:00:00 GMT",
        "notification_type": "email",
        "service_id": "REDACTED",
        "service_name": "Heartbeat sender"
    },
    {
        "count_delivered": 1,
        "count_permanent_failure": 0,
        "count_sending": 0,
        "count_sent": 0,
        "count_technical_failure": 0,
        "count_temporary_failure": 0,
        "date_created": "Sun, 01 Mar 2020 00:00:00 GMT",
        "notification_type": "email",
        "service_id": "REDACTED",
        "service_name": "Ioana's TBS Live Service"
    },
    // etc.
]
```

**k8s API**

```json
// Array of arrays
[
    [
        "Sun, 01 Mar 2020 00:00:00 GMT",
        "REDACTED",
        "Heartbeat sender",
        "email",
        0,
        22075,
        0,
        30,
        155,
        0
    ],
    [
        "Sun, 01 Mar 2020 00:00:00 GMT",
        "REDACTED",
        "Ioana's TBS Live Service",
        "email",
        0,
        1,
        0,
        0,
        0,
        0
    ]
    // etc...
]
```

## SQLAlchemy dependency bump

To add to the mystery, when we switched to the latest SQLAlchemy library version (from 1.3 to 1.4), the returned data type in Kubernetes environment switched from a `list` to `dict` datatype. When running locally though, testing from one developer showed that the returned data type would still be `list`.

**Comment by patheard** [comment]
@jimleroyer tracked this down to the following line in Admin, because of the different format of the JSON response:
https://github.com/cds-snc/notification-admin/blob/d0f439a444cc452c9aa26f43c9287d851193577f/app/main/views/platform_admin.py#L349

**Comment by yaelberger-commits** [comment]
Needs some investigation to give proper estimate

**Comment by yaelberger-commits** [comment]
Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Notify-Planning-614b3ad91bc2030015ed22f5/issues/cds-snc/notification-planning/653?planning-poker) with ZenHub @jimleroyer

**Comment by sastels** [comment]
The long term fix is to ensure that api returns data in a consistent format (preferably using dicts rather than lists where possible).

**Comment by adriannelee** [comment]
Jumana will try to replicate the issue on her computer today

**Comment by jzbahrai** [comment]
I ran this 
```
fields @timestamp, @message
| filter @message like /rror/
| filter @message not like /Telemetry client error:/
| filter @message not like /SAWarning: relationship 'ScheduledNotification.notification'/
| filter @message not like /UserWarning: The name 'v2/
| filter @message not like /email_address Not a valid email address/
| filter @message not like /team-only/
| sort @timestamp desc
```

for api-lambda logs for the past 2 weeks and didn't see any similar errors to the ones stated above - returning dicts or lists

**Comment by adriannelee** [comment]
Jumana will wrap this card up - end of time boxing

**Comment by jzbahrai** [comment]
This is what I did:

I pulled images from the ECR repository https://gallery.ecr.aws/v6b8u5o6/notify-api
I would then add .env files, and edit the code to make the monthly-data-by-service call locally

I found that the change occurred from lists to dict - 63cccb2 - but there was a json encoding error (not related to the above)
and it was fixed by 186f1a1. I couldn't pinpoint the exact change/ package this occurred in.

I am going to leave this for now, and instead revert the code Steve made to complete this ticket .
I did this overtime to see where the change occurred.
